New game database layout

Cari1111 opened the pull request:
## Issues

- resolves #53 

## General Changes

- added `game_database_test.py` in `tests`. This test should only enabled for manual testing because it inserts dummy values into the database
- changed file name into `game_database.py`
- changed db from object to module
- added debug log, when new game is inserted


## New Database Layout

- `game_ID`: Is automatically generated (increased) when new row is inserted
- `player1`: ID of user 1
- `player2`: ID of user 2
- `score1`: score of user 1
- `score2`: score of user 2
- `start_time`: start time of the game
- `end_state`: WIN, DRAW, DISCONNECTED
- `winner`: user ID of the winner
- `disconnected`: user ID of the disconected player

## New Methods

I added specific insert methodes which all use the `_insert_game()` methode:

- `insert_won_game` to insert a game that was won
- `insert_disconnected_game` to insert a game where players disconnected
- `insert_drawn_game` to insert a game that was drawn

I added methodes that use the SQLite function `COUNT()` to count things for diffrent statistics:

- `count_played_games`: count all games the user has played (disconnected and drawn games included)
- `count_won_games`: count all games where the user has won
- `count_disconnected_games`: count all games where the user disconnected
- `count_games_with_disconnect`: count all games where the user or the opponent disconnected
- `win_rate`: calculates the win-rate of the user (ignoring disconnected games)

## Still TODO

- [x]  rework naming
- [x] _add count methodes_
- [x]  _ID with _ or not?_: We agreed on `[...]_id`
- [x]  _update docstrings_

PacoSchatz commented:
Thanks for your implementation. I had a look at it and tested it locally - it works perfectly. One little thing I found:

in 
https://github.com/martius-lab/teamproject-competition-server/blob/962fc6a294fa104f56bf774592584f603c8c638a/teamprojekt_competition_server/server/databases.py#L70
 you still print the current time, each time an entry is written. Maybe you could convert this to a log message with a few more details, on what exactly was added to the database.

Apart from that it looks good to me. I like, that you added `insert_won_game()` and `insert_disconnected_game()`. That makes it easier to understand what game you are adding. 
But I can't say much, about how often we will use the functions that use the SQLite function `COUNT()` and we will see if any of these are unnecessary or if we might want to add another

Raining-Cloud commented on teamprojekt_competition_server/server/main.py:
```diff
@@ -54,12 +53,29 @@ def main():
-    gameID = game_db.insert_game(12, 23)
-    gameID = game_db.insert_game(52, 12)
-    print(gameID)
-    print(game_db.get_playerIDs(gameID=gameID))
-    print(game_db.get_gameIDs(playerID=12))
-    print(game_db.get_won_gameIDs(playerID=12))
+    gameID1 = game_db.insert_won_game(
+        winner_id=23, loser_id=4, score_winner=3, score_loser=6
+    )
+    gameID2 = game_db.insert_won_game(
+        winner_id=43,
+        loser_id=23,
+        score_winner=6,
+        score_loser=7,
+    )
+    gameID3 = game_db.insert_disconnected_game(
+        disconnected_user_id=1,
+        other_user_id=23,
+        score_disconnected_user=6,
+        score_other_user=7,
+    )
+    print(gameID1, gameID2, gameID3)
+    print(game_db.get_user_ids(game_id=gameID2))
+    print(game_db.get_game_ids(user_id=23))
+    print(game_db.get_won_game_ids(user_id=23))
+    print(game_db.count_played_games(user_id=23))
+    print(game_db.count_won_games(user_id=23))
+    print(game_db.count_disconnected_games(user_id=23))
+    print(game_db.win_rate(user_id=23))
```
there is a test folder btw. don't use main for this

Cari1111 replied:
added these tests to a seperate file in this folder

Raining-Cloud commented on teamprojekt_competition_server/server/databases.py:
```diff
@@ -1,8 +1,11 @@
+from enum import Enum
+from datetime import datetime
 
 GAME_DB_NAME = "game"
```
I am still confused by this, we fix the name of the database but need to create a new object for acsessing the same database everytime ?

Cari1111 replied:
changed it into a module

Raining-Cloud reviewed (changes requested):
I just took a quick look and found some things :)

Cari1111 commented on teamprojekt_competition_server/server/databases.py:
```diff
@@ -18,69 +21,291 @@ def __init__(self) -> None:
+        Returns:
+            int | None: game ID
+        """
+        return self._insert_game(
+            user1_id=disconnected_user_id,
+            user2_id=other_user_id,
+            score_user_1=score_disconnected_user,
+            score_user_2=score_other_user,
+            start_time=start_time,
+            game_end_state=GameEndState.DISCONNECTED.value,
+        )
+
+    def insert_drawn_game(
+        self,
+        user1_id: int,
+        user2_id: int,
+        score_user_1: float,
+        score_user_2: float,
+        start_time=None,
+    ) -> int | None:
+        """insert a drawn game into the database
+
+        Args:
+            user1_id (int): user ID of user 1
+            user2_id (int): user ID of user 2
+            score_user_1 (float): score of user 1
+            score_user_2 (float): score of user 2
+            start_time (_type_, optional): use datetime module. Defaults to None.
+
+        Returns:
+            int | None: game ID
+        """
+        return self._insert_game(
+            user1_id=user1_id,
+            user2_id=user2_id,
+            score_user_1=score_user_1,
+            score_user_2=score_user_2,
+            start_time=start_time,
+            game_end_state=GameEndState.DRAW.value,
+        )
```
you can still use the `_insert_game()` methode. I discussed this with @PacoSchatz and we decided that this is more clear to use these wrapper methodes.

Raining-Cloud replied:
Well still unecceserry imo but if all other need this 🤷

Cari1111 commented on teamprojekt_competition_server/server/databases.py:
```diff
@@ -18,69 +21,291 @@ def __init__(self) -> None:
+
+        disconnected_id = None
+        if game_end_state == GameEndState.DISCONNECTED.value:
+            disconnected_id = user1_id if is_user1_disconnected else user2_id
```
Yes

Cari1111 commented:
@PacoSchatz 
> Thanks for your implementation. I had a look at it and tested it locally - it works perfectly. One little thing I found:
> 
> in
> 
> https://github.com/martius-lab/teamproject-competition-server/blob/962fc6a294fa104f56bf774592584f603c8c638a/teamprojekt_competition_server/server/databases.py#L70
> 
> you still print the current time, each time an entry is written. Maybe you could convert this to a log message with a few more details, on what exactly was added to the database.
> 
> Apart from that it looks good to me. I like, that you added `insert_won_game()` and `insert_disconnected_game()`. That makes it easier to understand what game you are adding. But I can't say much, about how often we will use the functions that use the SQLite function `COUNT()` and we will see if any of these are unnecessary or if we might want to add another

Thanks for the review. I removed the print and added a log message.

Cari1111 commented on teamprojekt_competition_server/server/databases.py:
```diff
@@ -18,69 +21,291 @@ def __init__(self) -> None:
-        self.cursor.execute(f"INSERT INTO game VALUES ({winnerID}, {loserID})")
+        if start_time is None:
+            start_time = datetime.now()
+            print(start_time)
```
See response to paco.
